src/record/json-path.js

```diff
@@ -49,58 +32,74 @@ JsonPath.prototype.getValue = function() {
  * @param {Mixed} value
  *
  * @public
- * @returns {void}
+ * @returns updated value
  */
-JsonPath.prototype.setValue = function( value ) {
-	var node = this._record._$data,
-		i;
+module.exports.set = function( data, path, value, deepCopy ) {
+	var tokens = tokenize( path );
+
+	if ( deepCopy !== false ) {
+		value = utils.deepCopy( value );
+	}
 
-	for( i = 0; i < this._tokens.length - 1; i++ ) {
-		if( node[ this._tokens[ i ] ] !== undefined ) {
-			node = node[ this._tokens[ i ] ];
+	if ( tokens.length === 0 ) {
+		return value;
+	}
+
+	data = utils.shallowCopy( data );
+
+	var node = data;
+	for( var i = 0; i < tokens.length; i++ ) {
+		if ( i === tokens.length - 1) {
+			node[ tokens[ i ] ] = value;
 		}
-		else if( this._tokens[ i + 1 ] && !isNaN( this._tokens[ i + 1 ] ) ){
-			node = node[ this._tokens[ i ] ] = [];
+		else if( node[ tokens[ i ] ] !== undefined ) {
+			node = node[ tokens[ i ] ] = utils.shallowCopy( node[ tokens[ i ] ] );
+		}
+		else if( tokens[ i + 1 ] && !isNaN( tokens[ i + 1 ] ) ){
+			node = node[ tokens[ i ] ] = [];
 		}
 		else {
-			node = node[ this._tokens[ i ] ] = {};
+			node = node[ tokens[ i ] ] = Object.create( null );
 		}
 	}
 
-	node[ this._tokens[ i ] ] = value;
+	return data;
 };
 
 /**
  * Parses the path. Splits it into
  * keys for objects and indices for arrays.
  *
- * @private
- * @returns {void}
+ * @returns Array of tokens
  */
-JsonPath.prototype._tokenize = function() {
-	var parts = this._path.split( SPLIT_REG_EXP ),
-		part,
-		i;
+function tokenize( path ) {
+	if ( cache[ path ] ) {
+		return cache[ path ];
+	}
 
-	for( i = 0; i < parts.length; i++ ) {
```

Comment:
@yasserf: Since the tokens are cached. Maybe we can afford some validation here and throw if the path is invalid? Instead of split we could use a match.